apps/ejabberd/src/ejabberd_c2s.erl

```diff
@@ -790,6 +774,51 @@ wait_for_session_or_sm({xmlstreamerror, _}, StateData) ->
 wait_for_session_or_sm(closed, StateData) ->
     {stop, normal, StateData}.
 
+maybe_do_compress(El, NextState, StateData) ->
+    SockMod = (StateData#state.sockmod):get_sockmod(StateData#state.socket),
+    {Zlib, _} = StateData#state.zlib,
+    #xmlel{name = Name, attrs = Attrs, children = _} = El,
+    case {xml:get_attr_s(<<"xmlns">>, Attrs), Name} of
+        {?NS_COMPRESS_BIN, <<"compress">>} when Zlib == true,
+                                                ((SockMod == gen_tcp) or
+                                                 (SockMod == ejabberd_tls)) ->
+            do_compress(El, NextState, StateData);
+        _ ->
+            process_unauthenticated_stanza(StateData, El),
+            fsm_next_state(NextState, StateData)
+
+    end.
+
+do_compress(El, NextState, StateData) ->
+    {_, ZlibLimit} = StateData#state.zlib,
+    Auth = StateData#state.authenticated,
+    case Auth of
+        false ->
+            send_element(StateData, compress_setup_failed()),
+            fsm_next_state(NextState, StateData);
+        _ ->
+            case xml:get_subtag(El, <<"method">>) of
+                false ->
+                    send_element(StateData, compress_setup_failed()),
+                    fsm_next_state(NextState, StateData);
+                Method ->
+                    case xml:get_tag_cdata(Method) of
```

Comment:
This case is too deep. I think new function here instead of case would be more readable.